### ISSUE TITLE
[Bug Report] inconsistent global src and local rank in dist.nn.functional.broadcast

### ISSUE THREAD

**Issue by airsplay** [pytorch/pytorch#130515]
In torch.distributed.nn.functional._Broadcast, the below lines check whether the src and the rank are the same:

https://github.com/pytorch/pytorch/blob/dfd1d1971ea1265c597124b7e75fe5c8dd5a45b4/torch/distributed/nn/functional.py#L241-L242

However, the src here is global rank in the default process group WORLD (required by dist.broadcast), while the ctx.rank 

https://github.com/pytorch/pytorch/blob/dfd1d1971ea1265c597124b7e75fe5c8dd5a45b4/torch/distributed/nn/functional.py#L231

is a local rank under the provided process group by the function argument.

Thus this line fails to check whether the rank is the src, which results in all gradient to be zero for subgroups.

**Comment by LucasLLC** [comment]
c @wconstab

**Comment by wconstab** [comment]
a few things

- we want to deprecate these nn.functional apis since we have added other functional APIs.  
- i agree with the bug report- it would be better to change the nn.functional collectives to save the global rank instead of the group-local rank to match the behavior of other c10d collectives
- there is a long standing nit that the c10d apis are annoyingly implemented to use global ranks even when a processgroup is passed in.  we would like to change these to use group-local ranks, but this is a backward-compatibility breaking change, so i'm not sure if it will happen.

would you be willing to submit a PR to fix the nn.functional api? @airsplay

**Comment by kunwar-vikrant** [comment]
Hi! I’d like to work on this. I plan to add a subgroup regression test for the backward pass of `torch.distributed.nn.functional.broadcast` and fix the global-vs-group rank comparison. I’ll keep the change scoped and hold the PR until the issue is marked `actionable`.
